gui/util/range_slider.py

```diff
@@ -0,0 +1,294 @@
+"""
+
+from PyQt5 import QtCore, QtGui, QtWidgets
+
+
+class QRangeSlider(QtWidgets.QWidget):
+    """
+    QRangeSlider class, super class for QVRangeSlider and QHRangeSlider.
+    """
+
+    rangeChanged = QtCore.pyqtSignal(float, float)
+
+    def __init__(self, slider_range, values, parent=None):
+        QtWidgets.QWidget.__init__(self, parent)
+        self.bar_width = 4
+        self.emit_while_moving = 0
+        self.moving = "none"
+        self.old_scale_min = 0.0
+        self.old_scale_max = 0.0
+        self.scale = 0
+        self.setMouseTracking(False)
+        self.single_step = 0.0
+
+        self.collapsed = False
+        self.prev_moving = None
+        self.bc_min = None
+        self.bc_max = None
+
+        # Variables initialized in methods
+        self.scale_min = None
+        self.scale_max = None
+        self.start_display_min = None
+        self.start_display_max = None
+        self.start_pos = None
+        self.display_min = None
+        self.display_max = None
+
+        if slider_range:
+            self.setRange(slider_range)
+        else:
+            self.setRange([0.0, 1.0, 0.01])
+        if values:
+            self.values = values
+        else:
+            self.values = [0.3, 0.6]
+
+    def emitRange(self):
+        if (self.old_scale_min != self.scale_min) or (self.old_scale_max != self.scale_max):
+            self.rangeChanged.emit(self.scale_min, self.scale_max)
+            self.old_scale_min = self.scale_min
+            self.old_scale_max = self.scale_max
+            # For debug purposes
+            # if False:
+            #     print("Range change:", self.scale_min, self.scale_max)
+
+    @property
+    def values(self):
+        """
+        Values of the range bar
+        :return: [start, end]
+        """
+        return [self.scale_min, self.scale_max]
+
+    @values.setter
+    def values(self, values):
+        """
+        Setter for values of the range bar
+        :param values: [start, end]
+        """
+        self.scale_min, self.scale_max = values
+        self.emitRange()
+        self.updateDisplayValues()
+        self.update()
+
+    def mouseMoveEvent(self, event):
+        size = self.rangeSliderSize()
+        diff = self.start_pos - self.getPos(event)
+        if self.moving == "min":
+            temp = self.start_display_min - diff
+            if (temp >= self.bar_width) and (temp <= self.start_display_max):
+                self.display_min = temp
+                if self.display_max < self.display_min:
+                    self.display_max = self.display_min
+        elif self.moving == "max":
+            temp = self.start_display_max - diff
+            if (temp >= self.start_display_min) and (temp < size - self.bar_width):
+                self.display_max = temp
+                if self.display_max < self.display_min:
+                    self.display_min = self.display_max
+        elif self.moving == "bar":
```

Comment:
This sort "elif cascade" makes me wish Python had switch statements. In the meantime though I tend to use dicts of methods:

```
self.update_on_mouse_move[self.currently_moving](diff)
```

Incidentally I think `currently_moving`, although more verbose, reads more "obviously" than `self.moving`.

Reply:
another option is `ismoving` or `is_moving`

Reply:
Well, `is_moving` to me suggests that the widget itself is moving.... `self.component_moving`?